File: Python/DFS/PathWithGivenSequence.py
# @author
# Aakash Verma

# Output:
# true

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class PathWithGivenSequence:
    
    def __init__(self): 
        self.root = None

    def findSequence(self, root, sequence, index):
    	if root is None: return False

    	if index >= len(sequence) or root.data != sequence[index]:
    		return False

    	if(root.left is None and root.right is None and index == len(sequence) - 1):
    		return True

    	return self.findSequence(root.left, sequence, index + 1) or self.findSequence(root.right, sequence, index + 1)

    def hasPath(self, root, sequence):
        if root is None: return len(sequence) == 0

        return self.findSequence(root, sequence, 0)

        
        
if __name__ == '__main__':
    tree = PathWithGivenSequence()
    tree.root = Node(3)
    tree.root.left = Node(7)
    tree.root.right = Node(1)
    tree.root.left.left = Node(9)
    tree.root.right.left = Node(2)
    tree.root.right.right = Node(5)
    ans = tree.hasPath(tree.root, [3, 1, 2])
    print(ans)

File: Java/DFS/PathWithGivenSequence.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   true
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class PathWithGivenSequence {

    static Node root;
    PathWithGivenSequence() {
        root = null;
    }

    private static boolean findSequence(Node root, int[] sequence, int index) {
        if(root == null) return false;

        if(index >= sequence.length || root.data != sequence[index]) {
            return false;
        }

        if(root.left == null && root.right == null && index == sequence.length - 1) {
            return true;
        }

        return findSequence(root.left, sequence, index + 1) || findSequence(root.right, sequence, index + 1);
    }

    public static boolean hasPath(Node root, int[] sequence) {
        if(root == null) return sequence.length == 0;

        return findSequence(root, sequence, 0);
    }

    public static void main (String[] args) {
        root = new Node(3);
        root.left = new Node(7);
        root.right = new Node(1);
        root.left.left = new Node(9);
        root.right.left = new Node(2);
        root.right.right = new Node(5);
        boolean ans = hasPath(root, new int[] {3, 1, 2});
        System.out.println(ans);
    }
}